Feature: use token informations (#1585)
stellio-hub/stellio-context-broker: docs/admin/upgrading_to_2.31.0.md

```
APPLICATION_SEARCH_SERVICE_URL=http://my-hostname:8083
APPLICATION_SUBSCRIPTION_SERVICE_URL=http://my-hostname:8084
````
If you don't use one of these variables, the change will not impact you.

> ## Authorization based on JSON Web Token (JWT)

Reviewer: What bothers me here is that there is a presentation of a new feature and a migration procedure. The migration procedure belongs to this page, the new feature should be introduced in authentication_integration.md.

Author: The `authentication_integration.md` already describe the feature. But here it describe what change with authentication in the new version. I can shorten it and put a link to `authentication_integration.md` if you prefer

Reviewer: yes, better soc.

The new authorization system works using only the OIDC token. Meaning you can use Stellio authorization features with any OIDC provider.
You can configure what JWT claims are considered to evaluate access rights with the `application.authentication.claims-paths` environment variable.
```
application.authentication.claims-paths = realm_access.roles,groups_uuids
```

> Stellio retrieves all the configured claims in the user token (as well as the user sub) and uses all permissions assigned to one of the claims to evaluate the user rights.

Reviewer: ```suggestion
Stellio retrieves all the configured claims in the user token (as well as the user sub) and uses the permissions assigned to the claims to evaluate the user rights.
```

> ## Migrate the current authorization setup

Reviewer: It should be a sub-section of "Migrate the current authorization setup" (and this will make the "out of the box, this lets you ..." understandable)

Out of the box, this lets you assign permission to Keycloak roles instead of groups.
It also means that a desynchronization of Stellio subjects information will no longer impact the NGSI-LD endpoints. (only the subject endpoints)

### Migrate groups permission
> **Warning:** You should follow this migration if you have permissions targeting groups.

Existing permission targeting groups need to access the user groups ids in the token.

> For this we have developed a new token mapper which is present in the Keycloak images provided by us starting from [version 26.5.5)(https://hub.docker.com/repository/docker/easyglobalmarket/keycloak/tags/26.5.5/sha256-9746311b62a0300b5834bbea1d10300c0977caf5da22dcb790d75a58f403a6a7).

Reviewer: ```suggestion
For this we have developed a new token mapper which is present in the Keycloak images provided by us starting from [version 26.5.5](https://hub.docker.com/repository/docker/easyglobalmarket/keycloak/tags/26.5.5/sha256-9746311b62a0300b5834bbea1d10300c0977caf5da22dcb790d75a58f403a6a7).
```

Once the keycloak image is upgraded, you can configure the token mapper to add the groups uuids in the token.

#### Add the Groups UUID Mapper in Clients scopes > roles > Mappers > Add Mapper > by configuration 
You can use the `roles` scope or create your own scope (make sure it is used when generating the token).
![](images/group-uuid-mapper-configuration/step-1.png)

#### Configure the mapper to add the ids behind the `groups_uuids` claim.
![](images/group-uuid-mapper-configuration/step-2.png)

#### Verify that the groups uuids are present in the token. (in Clients > your-client > Clients scopes > evaluate > Generated access token)
![](images/group-uuid-mapper-configuration/step-3.png)

> When all the realms used by stellio have the `groups_uuids` claim configured. You are ready to upgrade to version 2.31.0.

Reviewer: ```suggestion
When all the realms used by Stellio have the `groups_uuids` claim configured, you are ready to upgrade to version 2.31.0.
```

```
APPLICATION_SEARCH_SERVICE_URL=http://my-hostname:8083
APPLICATION_SUBSCRIPTION_SERVICE_URL=http://my-hostname:8084
````
If you don't use one of these variables, the change will not impact you.

> ## Authorization based on JWT token

Reviewer: you know the 'T' in JWT is for token ;)

you could rewrite it "Authorization based on JSON Web Tokens (JWT)"

> The new authorization system work using only the OIDC token. Meaning you can use Stellio authorization features with any OpenId-Connect provider.

Reviewer: ```suggestion
The new authorization system works using only the OIDC token. Meaning you can use Stellio authorization features with any OIDC provider.
```

> You can configure what JWT claims are considered with the `application.authentication.claims-paths` environment variable.

Reviewer: ```suggestion
You can configure what JWT claims are considered to evaluate access rights with the `application.authentication.claims-paths` environment variable.
```

```
application.authentication.claims-paths = realm_access.roles,groups_uuids
```
Stellio checks all the configured claims in the user token (+ the user sub).

> And use all permissions assigned to one of the user claims to evaluate the user rights.

Reviewer: ```suggestion
Stellio retrieves all the configured claims in the user token (as well as the user sub) and uses all permissions assigned to one of the claims to evaluate the user rights.
```

## Migrating the current authorization setup

> Out of the box, this will let you assign permission to Keycloak roles instead of groups.

Reviewer: ```suggestion
Out of the box, this lets you assign permission to Keycloak roles instead of groups.
```

It also means that a desynchronization of Stellio subjects information will no longer impact the NGSI-LD endpoints. (only the subject endpoints)

> ### Migrate groups permission

Reviewer: you say "Migrating" above and "Migrate" here. Be consistent.

> **Warning:** You should follow this migration if you have permissions targeting groups.

> Existing permission targeting groups id need to access the user groups ids in the token.

Reviewer: ```suggestion
Existing permission targeting groups need to access the user groups ids in the token.
```

> For this we have developed a new token mapper which is present in the new keycloak image (easyglobalmarket/keycloak:26.5.5).

Reviewer: ```suggestion
For this we have developed a new token mapper which is present in the Keycloak images provided by us starting from [version 26.5.5)(https://hub.docker.com/repository/docker/easyglobalmarket/keycloak/tags/26.5.5/sha256-9746311b62a0300b5834bbea1d10300c0977caf5da22dcb790d75a58f403a6a7).
```

Once the keycloak image is upgraded, you can configure the token mapper to add the groups uuids in the token.

#### Select the Groups UUID Mapper in Clients scopes > my-client-scope > Mappers > Add Mapper > by configuration 

> (You can use an existing client scope or create your own, make sure it is used when generating the token)

Reviewer: not really clear what this client scope is about. (and what are the impacts for all existing clients)

Author: `Client scope` is the menu option. I changed the description to avoid confusion between `Client Scope` in the menu and the `scope` inside it.

Reviewer: Sorry, my comment was not clear. My point was about the "my-scope" scope.

![](images/group-uuid-mapper-configuration/step-1.png)

#### Configure the mapper to add the ids behind the `groups_uuids` claim.
![](images/group-uuid-mapper-configuration/step-2.png)

#### Verify that the groups uuids are present in the token. (in Clients > your-client > Clients scopes > evaluate > Generated access token)
![](images/group-uuid-mapper-configuration/step-3.png)

> When all the realms used by stellio have the `groups_uuids` claim configured. You are ready to upgrade to stellio:2.31.0.

Reviewer: ```suggestion
When all the realms used by stellio have the `groups_uuids` claim configured. You are ready to upgrade to version 2.31.0.
```
